src/main/java/com/google/sps/servlets/SearchServlet.java

```diff
@@ -55,64 +58,116 @@
+      );
+      request.setAttribute("recipesList", recipesListToReturn);
+    } catch (SearchQueryException e) {
+      response.setStatus(505);
+    }
     request.getRequestDispatcher("/search.jsp").forward(request, response);
   }
 
   /**
    * Replace and return the string with no commas and no multiple consecutive or trailing spaces.
    */
-  private String trimAndRemoveCommas(String stringToRemove) {
-    if (!StringUtils.isBlank(stringToRemove)) {
-      stringToRemove = stringToRemove.replaceAll("[,]+", " ");
-      stringToRemove = stringToRemove.replaceAll("  +", " ");
+  private String sanitizeString(String stringToSanitize) {
+    if (!StringUtils.isBlank(stringToSanitize)) {
+      stringToSanitize = stringToSanitize.replaceAll("(,\\s?)+", " ");
+    }
+    return stringToSanitize;
+  }
+
+  /**
+   * If the fields exists, returns the parameter from the request, otherwise it returns an empty string.
+   */
+  private String getParameter(String field, HttpServletRequest request) {
+    String parameter = new String();
+    try {
+      parameter = request.getParameter(field);
+    } catch (NullPointerException e) {
+      parameter = "";
     }
-    return stringToRemove;
+    return parameter;
   }
 
   /**
-   * Returns a list of recipes matching the given title and/or ingredients.
+   * Returns a list of recipes matching the given title and ingredients.
    */
   private List<Recipe> recipesMatching(
     HttpServletRequest request,
     HttpServletResponse response,
-    String stringToMatch
+    String stringToMatch,
+    String difficulty,
+    Integer time
   )
     throws ServletException, IOException {
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    List<Recipe> recipesMatchingList = new ArrayList<>();
     Index index = getIndex("recipes_index");
-
-    Query query = buildQuery(stringToMatch, 10);
+    Query query = buildQuery(stringToMatch, difficulty, time);
     Results<ScoredDocument> results = index.search(query);
-    for (ScoredDocument entity : results) {
+    List<Recipe> matchingRecipes = new ArrayList<>();
+    for (ScoredDocument document : results) {
       try {
-        Entity recipeEntity = getRecipeEntityFromDocumentEntity(
-          datastore,
-          entity
+        Entity recipeEntity = getRecipeEntityFromDocument(datastore, document);
+        ArrayList<String> matchingIngredients = returnMatchingIngredients(
+          document
         );
-
-        Recipe recipe = buildRecipe(recipeEntity, datastore);
-        recipesMatchingList.add(recipe);
+        Recipe recipe = buildRecipe(recipeEntity, matchingIngredients);
+        matchingRecipes.add(recipe);
       } catch (EntityNotFoundException e) {
         response.setStatus(505);
+        request.setAttribute("recipesList", new ArrayList<>());
         request.getRequestDispatcher("/search.jsp").forward(request, response);
-        return new ArrayList<>();
+        return matchingRecipes;
```

Comment:
Leaving the same comment here for the second time (for some reason this was changed again!):

You should return an empty list in this case.

Reply:
I changed by mistake when merging the commits.